akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PerformanceSpec.scala

```diff
@@ -0,0 +1,138 @@
+          case command          ⇒ other(command, parameters)
+        },
+        eventHandler = {
+          case (state, _) ⇒ state
+        }
+      ).onRecoveryCompleted {
+          case (_, _) ⇒ if (parameters.every(1000)) print("r")
+        }
+    }).onFailure(SupervisorStrategy.resume)
+  }
+
+  def eventSourcedTestPersistenceBehavior(name: String, probe: TestProbe[Command]) =
+    behavior(name, probe) {
+      case (CommandWithEvent(evt), parameters) ⇒
+        Effect.persist(evt).andThen({
+          parameters.persistCalls += 1
+          if (parameters.every(1000)) print(".")
+          if (parameters.shouldFail) throw TE("boom")
+        })
+    }
+}
+
+class PerformanceSpec extends ActorTestKit with TypedAkkaSpecWithShutdown with Eventually {
+
+  import PerformanceSpec._
+
+  val loadCycles = system.settings.config.getInt("akka.persistence.performance.cycles.load")
+
+  override def config: Config =
+    ConfigFactory.parseString(
+      s"""
+      akka.actor.serialize-creators = off
+      akka.actor.serialize-messages = off
+      akka.actor.warn-about-java-serializer-usage = off
+      akka.persistence.publish-plugin-commands = on
+      akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
+      akka.persistence.journal.leveldb.dir = "target/journal-PerformanceSpec"
+      akka.persistence.snapshot-store.plugin = "akka.persistence.snapshot-store.local"
+      akka.persistence.snapshot-store.local.dir = "target/snapshots-PerformanceSpec/"
+      akka.test.single-expect-default = 10s
+    """).withFallback(ConfigFactory.parseString(PerformanceSpec.config))
+
+  def stressPersistentActor(persistentActor: ActorRef[Command], probe: TestProbe[Command],
+                            failAt: Option[Long], description: String): Unit = {
+    failAt foreach { persistentActor ! FailAt(_) }
+    val m = new Measure(loadCycles)
+    m.startMeasure()
+    1 to loadCycles foreach { i ⇒ persistentActor ! CommandWithEvent(s"msg$i") }
+    persistentActor ! StopMeasure
+    probe.expectMessage(100.seconds, StopMeasure)
+    println(f"\nthroughput = ${m.stopMeasure()}%.2f $description per second")
+  }
+
+  def stressEventSourcedPersistentActor(failAt: Option[Long]): Unit = {
+    val probe = TestProbe[Command]
+    val name = s"${this.getClass.getSimpleName}-${UUID.randomUUID().toString}"
+    val persistentActor = spawn(eventSourcedTestPersistenceBehavior(name, probe), name)
+    stressPersistentActor(persistentActor, probe, failAt, "persistent events")
+  }
+
+  "An event sourced persistent actor" should {
+    "have some reasonable throughput" in {
+      stressEventSourcedPersistentActor(None)
+    }
+    //    FIXME: PersistentActor survives exception but persistent behavior does not, losing pending messages
+    //    "have some reasonable throughput under failure conditions" in {
+    //      stressEventSourcedPersistentActor(Some(loadCycles / 10))
+    //    }
```

Comment:
I think the reason is that typed actors are stopped by default when exception is thrown and untyped are restarted (keeping same mailbox). You can try to add `Behaviors.supervise` with `SupervisorStrategy.restart`.

Ah, you already tried with `resume`.

When running it with `restart` and enabling the stack trace for the `TE` I see this:
```
[ERROR] [03/26/2018 14:28:33.045] [PerformanceSpec-akka.actor.default-dispatcher-4] [akka://PerformanceSpec/user/PerformanceSpec-44f0b7fe-ef2d-4a7c-9c1c-f9ad88fcc2c8][persistenceId:PerformanceSpec-44f0b7fe-ef2d-4a7c-9c1c-f9ad88fcc2c8][phase:persist-evts] boom
akka.testkit.typed.TE: boom
	at akka.persistence.typed.scaladsl.PerformanceSpec$$anonfun$eventSourcedTestPersistenceBehavior$1$$anonfun$apply$2.apply$mcV$sp(PerformanceSpec.scala:88)
	at akka.persistence.typed.scaladsl.Effect$$anonfun$andThen$1.apply(Effect.scala:73)
	at akka.persistence.typed.scaladsl.Effect$$anonfun$andThen$1.apply(Effect.scala:73)
	at akka.persistence.typed.internal.EventsourcedRunning.applySideEffect(EventsourcedRunning.scala:305)
	at akka.persistence.typed.internal.EventsourcedRunning.applySideEffects(EventsourcedRunning.scala:292)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onWriteResponse$1(EventsourcedRunning.scala:209)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onJournalResponse(EventsourcedRunning.scala:216)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onMessage(EventsourcedRunning.scala:183)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onMessage(EventsourcedRunning.scala:171)
	at akka.actor.typed.scaladsl.MutableBehavior.receive(MutableBehavior.scala:23)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:173)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Widened.receive(BehaviorImpl.scala:52)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.Supervisor.receive(Restarter.scala:96)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.adapter.ActorAdapter.akka$actor$typed$internal$adapter$ActorAdapter$$handleMessage(ActorAdapter.scala:64)
	at akka.actor.typed.internal.adapter.ActorAdapter$$anonfun$running$1.applyOrElse(ActorAdapter.scala:60)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:517)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:19)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
	at akka.actor.ActorCell.invoke(ActorCell.scala:557)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I wonder what that recursion can be?

I have to look at this more later...

Reply:
ok, I created issue https://github.com/akka/akka/issues/24787 for the weird stack.

The problem you are seeing here though is that the `StashBuffer` is recreated for each restart and thereby dropping messages in the stash. I'm not sure how we should solve that?

The only place we could keep the stash is in the `PersistentBehaviorImpl` but that could potentially be used for multiple `spawn`, but in practice that might never be possible since there can only be one instance with the same `persistenceId`. On the other hand it can be spawned again after `stop` and then the stash would survive that also which is not desired. No that's not good.

Reply:
Second thought, we can clean the stash in a PostStop signal if we hold on to the `StashBuffer` in `PersistentBehaviorImpl`. Perhaps that is the only way.

@kelebra Do you understand what I mean with the `StashBuffer` here and could you try to implement that idea of only creating the buffer once, and then use the same for subsequent calls to `PersistentBehaviorImpl.apply`?

Reply:
@patriknw, thanks a lot for explanation. So far as I understand that we create new instance of `StashBuffer` in `EventsourcedSetup.apply` which in turn happens during `PersistentBehaviorImpl.apply`. If `PersistentBehaviorImpl.apply` is called one more time for resume and/or restart we need to keep stash in order not to loose commands of internal protocol.

So if I understand you right I should do the following:

* Move `internalStash` creation out of `EventsourcedSetup` to `PersistentBehaviorImpl`
* Create `EventsourcedSetup` in `PersistentBehaviorImpl.apply` using `internalStash` from `PersistentBehaviorImpl` (as per first step)

Will try to add changes to this PR today itself for your review.

Reply:
I wonder which behaviors out of four implementations should be impacted or to specify question which sub-behaviors even should be impacted?

Reply:
Introduced `akka.persistence.typed.internal.EventsourcedStashReferenceManagement` that holds reference and prevents stash from being recreated on `apply` and mixed it into `PersistentBehaviorImpl` for creation using inherited `stash`. Uncommented `"have some reasonable throughput under failure conditions"` that now passes.